Virtualbox guest fails with latest kernel. Missing ttm library.

**Describe the bug**

I am no longer able to build VirtualBox with guest additions when switching over to the lastest linux kernel: `boot.kernelPackages = pkgs.linuxPackages_latest;`. The error complains that `ttm` headers are missing/not found. This bug occurs consistently across all 3 of my machines.
I am still able to build VirtualBox without guest additions.

**To Reproduce**

Here is the relevant snippet of my `configuration.nix`
```nix
virtualisation.virtualbox = {
  host = {
    enable = true;
    enableExtensionPack = true;
  };
  guest.enable = true;
};
boot = {
  kernelPackages = pkgs.linuxPackages_latest;
};
nixpkgs.config.allowUnfree = true;
```

**Additional context**


Here is the relevant errors from using `sudo nixos-rebuild switch`
```sh
=== Building 'vboxsf' module ===
make[1]: Entering directory '/build/install/src/vboxguest-6.0.8/vboxsf'
make V= CONFIG_MODULE_SIG= -C /nix/store/f67gigrcnnhswswacch6rn1mx7r0r22d-linux-5.2.4-dev/lib/modules/5.2.4/build M=/build/install/src/vboxguest-6.0.8/vboxsf SRCROOT=/build/install/src/vboxguest-6.0.8/vboxsf -j4 modules
make[2]: Entering directory '/nix/store/f67gigrcnnhswswacch6rn1mx7r0r22d-linux-5.2.4-dev/lib/modules/5.2.4/build'
  CC [M]  /build/install/src/vboxguest-6.0.8/vboxsf/vfsmod.o
  CC [M]  /build/install/src/vboxguest-6.0.8/vboxsf/dirops.o
  CC [M]  /build/install/src/vboxguest-6.0.8/vboxsf/lnkops.o
  CC [M]  /build/install/src/vboxguest-6.0.8/vboxsf/regops.o
  CC [M]  /build/install/src/vboxguest-6.0.8/vboxsf/utils.o
  CC [M]  /build/install/src/vboxguest-6.0.8/vboxsf/VBoxGuestR0LibGenericRequest.o
  CC [M]  /build/install/src/vboxguest-6.0.8/vboxsf/VBoxGuestR0LibHGCM.o
  CC [M]  /build/install/src/vboxguest-6.0.8/vboxsf/VBoxGuestR0LibIdc.o
  CC [M]  /build/install/src/vboxguest-6.0.8/vboxsf/VBoxGuestR0LibIdc-unix.o
  CC [M]  /build/install/src/vboxguest-6.0.8/vboxsf/VBoxGuestR0LibInit.o
  CC [M]  /build/install/src/vboxguest-6.0.8/vboxsf/VBoxGuestR0LibPhysHeap.o
  CC [M]  /build/install/src/vboxguest-6.0.8/vboxsf/VBoxGuestR0LibSharedFolders.o
  LD [M]  /build/install/src/vboxguest-6.0.8/vboxsf/vboxsf.o
  Building modules, stage 2.
  MODPOST 1 modules
  CC      /build/install/src/vboxguest-6.0.8/vboxsf/vboxsf.mod.o
  LD [M]  /build/install/src/vboxguest-6.0.8/vboxsf/vboxsf.ko
make[2]: Leaving directory '/nix/store/f67gigrcnnhswswacch6rn1mx7r0r22d-linux-5.2.4-dev/lib/modules/5.2.4/build'
make[1]: Leaving directory '/build/install/src/vboxguest-6.0.8/vboxsf'

=== Building 'vboxvideo' module ===
make[1]: Entering directory '/build/install/src/vboxguest-6.0.8/vboxvideo'
make V= CONFIG_MODULE_SIG= -C /nix/store/f67gigrcnnhswswacch6rn1mx7r0r22d-linux-5.2.4-dev/lib/modules/5.2.4/build M=/build/install/src/vboxguest-6.0.8/vboxvideo SRCROOT=/build/install/src/vboxguest-6.0.8/vboxvideo -j4 modules
make[2]: Entering directory '/nix/store/f67gigrcnnhswswacch6rn1mx7r0r22d-linux-5.2.4-dev/lib/modules/5.2.4/build'
  CC [M]  /build/install/src/vboxguest-6.0.8/vboxvideo/hgsmi_base.o
  CC [M]  /build/install/src/vboxguest-6.0.8/vboxvideo/modesetting.o
  CC [M]  /build/install/src/vboxguest-6.0.8/vboxvideo/vbox_drv.o
  CC [M]  /build/install/src/vboxguest-6.0.8/vboxvideo/vbox_fb.o
  CC [M]  /build/install/src/vboxguest-6.0.8/vboxvideo/vbox_irq.o
/build/install/src/vboxguest-6.0.8/vboxvideo/vbox_fb.c: In function 'vboxfb_create':
/build/install/src/vboxguest-6.0.8/vboxvideo/vbox_fb.c:335:2: warning: implicit declaration of function 'drm_fb_helper_fill_fix'; did you mean 'drm_fb_helper_fill_info'? [-Wimplicit-function-declaration]
  drm_fb_helper_fill_fix(info, fb->pitches[0], fb->format->depth);
  ^~~~~~~~~~~~~~~~~~~~~~
  drm_fb_helper_fill_info
  CC [M]  /build/install/src/vboxguest-6.0.8/vboxvideo/vbox_main.o
/build/install/src/vboxguest-6.0.8/vboxvideo/vbox_fb.c:339:2: warning: implicit declaration of function 'drm_fb_helper_fill_var'; did you mean 'drm_fb_helper_fill_info'? [-Wimplicit-function-declaration]
  drm_fb_helper_fill_var(info, &fbdev->helper, sizes->fb_width,
  ^~~~~~~~~~~~~~~~~~~~~~
  drm_fb_helper_fill_info
  CC [M]  /build/install/src/vboxguest-6.0.8/vboxvideo/vbox_mode.o
  CC [M]  /build/install/src/vboxguest-6.0.8/vboxvideo/vbox_ttm.o
  CC [M]  /build/install/src/vboxguest-6.0.8/vboxvideo/vbva_base.o
/build/install/src/vboxguest-6.0.8/vboxvideo/vbox_ttm.c:32:10: fatal error: ttm/ttm_page_alloc.h: No such file or directory
 #include <ttm/ttm_page_alloc.h>
          ^~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[4]: *** [/nix/store/f67gigrcnnhswswacch6rn1mx7r0r22d-linux-5.2.4-dev/lib/modules/5.2.4/source/scripts/Makefile.build:279: /build/install/src/vboxguest-6.0.8/vboxvideo/vbox_ttm.o] Error 1
make[4]: *** Waiting for unfinished jobs....
make[3]: *** [/nix/store/f67gigrcnnhswswacch6rn1mx7r0r22d-linux-5.2.4-dev/lib/modules/5.2.4/source/Makefile:1595: _module_/build/install/src/vboxguest-6.0.8/vboxvideo] Error 2
make[2]: *** [/nix/store/f67gigrcnnhswswacch6rn1mx7r0r22d-linux-5.2.4-dev/lib/modules/5.2.4/source/Makefile:179: sub-make] Error 2
make[2]: Leaving directory '/nix/store/f67gigrcnnhswswacch6rn1mx7r0r22d-linux-5.2.4-dev/lib/modules/5.2.4/build'
make[1]: *** [/build/install/src/vboxguest-6.0.8/vboxvideo/Makefile-footer.gmk:111: vboxvideo] Error 2
make[1]: Leaving directory '/build/install/src/vboxguest-6.0.8/vboxvideo'
make: *** [Makefile:63: vboxvideo] Error 1
builder for '/nix/store/9985ahn3zw8cdxzh9782j04kvzyd9slg-VirtualBox-GuestAdditions-6.0.8-5.2.4.drv' failed with exit code 2
cannot build derivation '/nix/store/r7zby4a3rqhyn0h7cxsj18a09id6rz8p-kernel-modules.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/xh66icwcw87mkp8vpsj4rpim0kb0nzlh-linux-5.2.4-modules.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/gj77lwcb3ki23g79aki79yjnrar1q2vn-system-path.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/xyn3rv7sm32jzv5w0v6zdi8ibm8l7f0c-unit-virtualbox.service.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/8q9fhxd4k2c93cd6p4n0vl2k9idz090g-xsession-wrapper.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/q5mlg1prnyz8ap26ka73vm08xm13n6jz-nixos-system-angela-19.09pre186857.239fffc90d7.drv': 1 dependencies couldn't be built
error: build of '/nix/store/q5mlg1prnyz8ap26ka73vm08xm13n6jz-nixos-system-angela-19.09pre186857.239fffc90d7.drv' failed
```

**Metadata**
Nix Info:
```sh
 - system: `"x86_64-linux"`
 - host os: `Linux 5.2.4, NixOS, 19.09pre186857.239fffc90d7 (Loris)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.2.2`
 - channels(camus): `"home-manager, nextbookit, nixos-mio"`
 - channels(root): `"nixos-19.09pre186857.239fffc90d7, nixos-hardware, nixos-stable-18.09.2574.a7e559a5504"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

Maintainer information:
```yaml
option: virtualisation.virtualbox.guest.enable
```

I take that back - it's now failing at a later stage:

```
/build/install/src/vboxguest-6.0.10/vboxguest/r0drv/linux/mp-r0drv-linux.c: In function 'VBoxGuest_RTMpOnAll':
/build/install/src/vboxguest-6.0.10/vboxguest/r0drv/linux/mp-r0drv-linux.c:287:18: error: void value not ignored as it ought to be
         int rc = smp_call_function(rtmpLinuxAllWrapper, &Args, 0 /* wait */);
                  ^~~~~~~~~~~~~~~~~
/build/install/src/vboxguest-6.0.10/vboxguest/r0drv/linux/mp-r0drv-linux.c: In function 'VBoxGuest_RTMpOnOthers':
/build/install/src/vboxguest-6.0.10/vboxguest/r0drv/linux/mp-r0drv-linux.c:341:8: error: void value not ignored as it ought to be
     rc = smp_call_function(rtmpLinuxWrapper, &Args, 1 /* wait */);
        ^
  CC [M]  /build/install/src/vboxguest-6.0.10/vboxguest/generic/RTSemEventMultiWaitNoResume-2-ex-generic.o
  CC [M]  /build/install/src/vboxguest-6.0.10/vboxguest/generic/rtStrFormatKernelAddress-generic.o
  CC [M]  /build/install/src/vboxguest-6.0.10/vboxguest/generic/errvars-generic.o
/build/install/src/vboxguest-6.0.10/vboxguest/r0drv/linux/memuserkernel-r0drv-linux.o: warning: objtool: rtR0MemKernelCopyLnxWorker()+0x12: redundant CLD
make[4]: *** [/nix/store/xjz5wdr0pkiyln3gv0dz2nlbf7484wfs-linux-5.3.1-dev/lib/modules/5.3.1/source/scripts/Makefile.build:281: /build/install/src/vboxguest-6.0.10/vboxguest/r0drv/linux/mp-r0drv-linux.o] Error 1
make[4]: *** Waiting for unfinished jobs....
make[3]: *** [/nix/store/xjz5wdr0pkiyln3gv0dz2nlbf7484wfs-linux-5.3.1-dev/lib/modules/5.3.1/source/Makefile:1624: _module_/build/install/src/vboxguest-6.0.10/vboxguest] Error 2
make[2]: *** [/nix/store/xjz5wdr0pkiyln3gv0dz2nlbf7484wfs-linux-5.3.1-dev/lib/modules/5.3.1/source/Makefile:179: sub-make] Error 2
```

@wkral This looks like we're missing a patch similar to the one in https://github.com/NixOS/nixpkgs/pull/69323.

--- a/pkgs/applications/virtualization/virtualbox/guest-additions/default.nix
+++ b/pkgs/applications/virtualization/virtualbox/guest-additions/default.nix
@@ -1,5 +1,5 @@
 { stdenv, fetchurl, lib, patchelf, cdrkit, kernel, which, makeWrapper
-, zlib, xorg, dbus, virtualbox }:
+, zlib, xorg, dbus, virtualbox, dos2unix }:
 
 let
   version = virtualbox.version;
@@ -32,9 +32,6 @@ in stdenv.mkDerivation {
   KERN_DIR = "${kernel.dev}/lib/modules/${kernel.modDirVersion}/build";
   KERN_INCL = "${kernel.dev}/lib/modules/${kernel.modDirVersion}/source/include";
 
-  # If you add a patch you probably need this.
-  #patchFlags = [ "-p1" "-d" "install/src/vboxguest-${version}" ];
-
   hardeningDisable = [ "pic" ];
 
   NIX_CFLAGS_COMPILE = "-Wno-error=incompatible-pointer-types -Wno-error=implicit-function-declaration";
@@ -42,10 +39,17 @@ in stdenv.mkDerivation {
   nativeBuildInputs = [ patchelf makeWrapper ];
   buildInputs = [ cdrkit ] ++ kernel.moduleBuildDependencies;
 
-  postPatch = ''
+
+  prePatch = ''
     substituteInPlace src/vboxguest-${version}/vboxvideo/vbox_ttm.c \
       --replace "<ttm/" "<drm/ttm/"
+    ${dos2unix}/bin/dos2unix src/vboxguest-${version}/vboxguest/r0drv/linux/mp-r0drv-linux.c
   '';
+
+  patchFlags = [ "-p1" "-d" "src/vboxguest-${version}" ];
+  # Kernel 5.3 fix, should be fixed with VirtualBox 6.0.14
+  # https://www.virtualbox.org/ticket/18911
+  patches = [ ./kernel-5.3-fix.patch ];
 
   unpackPhase = ''
     ${if stdenv.hostPlatform.system == "i686-linux" || stdenv.hostPlatform.system == "x86_64-linux" then ''

--- a/pkgs/applications/virtualization/virtualbox/guest-additions/kernel-5.3-fix.patch
+++ b/pkgs/applications/virtualization/virtualbox/guest-additions/kernel-5.3-fix.patch
@@ -1,0 +1,50 @@
+--- a/vboxguest/r0drv/linux/mp-r0drv-linux.c
++++ a/vboxguest/r0drv/linux/mp-r0drv-linux.c
+@@ -283,12 +283,15 @@
+     if (RTCpuSetCount(&OnlineSet) > 1)
+     {
+         /* Fire the function on all other CPUs without waiting for completion. */
+-# if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 27)
++# if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 3, 0)
++        smp_call_function(rtmpLinuxAllWrapper, &Args, 0 /* wait */);
++# elif LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 27)
+         int rc = smp_call_function(rtmpLinuxAllWrapper, &Args, 0 /* wait */);
++        Assert(!rc); NOREF(rc);
+ # else
+         int rc = smp_call_function(rtmpLinuxAllWrapper, &Args, 0 /* retry */, 0 /* wait */);
+-# endif
+         Assert(!rc); NOREF(rc);
++# endif
+     }
+ #endif
+
+@@ -326,7 +329,6 @@
+ {
+ #ifdef CONFIG_SMP
+     IPRT_LINUX_SAVE_EFL_AC();
+-    int rc;
+     RTMPARGS Args;
+
+     RTTHREADPREEMPTSTATE PreemptState = RTTHREADPREEMPTSTATE_INITIALIZER;
+@@ -337,14 +339,17 @@
+     Args.cHits = 0;
+
+     RTThreadPreemptDisable(&PreemptState);
+-# if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 27)
+-    rc = smp_call_function(rtmpLinuxWrapper, &Args, 1 /* wait */);
++# if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 3, 0)
++    smp_call_function(rtmpLinuxWrapper, &Args, 1 /* wait */);
++# elif LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 27)
++    int rc = smp_call_function(rtmpLinuxWrapper, &Args, 1 /* wait */);
++    Assert(rc == 0); NOREF(rc);
+ # else /* older kernels */
+-    rc = smp_call_function(rtmpLinuxWrapper, &Args, 0 /* retry */, 1 /* wait */);
++    int rc = smp_call_function(rtmpLinuxWrapper, &Args, 0 /* retry */, 1 /* wait */);
++    Assert(rc == 0); NOREF(rc);
+ # endif /* older kernels */
+     RTThreadPreemptRestore(&PreemptState);
+
+-    Assert(rc == 0); NOREF(rc);
+     IPRT_LINUX_RESTORE_EFL_AC();
+ #else
+     RT_NOREF(pfnWorker, pvUser1, pvUser2);

